### Reading in data

In [164]:
import pandas as pd
brands = pd.read_json('brands.json',lines=True)
receipts = pd.read_json('receipts.json',lines=True)
users = pd.read_json('users.json',lines=True)
receipts

_id  bonusPointsEarned  \
0     {'$oid': '5ff1e1eb0a720f0523000575'}              500.0   
1     {'$oid': '5ff1e1bb0a720f052300056b'}              150.0   
2     {'$oid': '5ff1e1f10a720f052300057a'}                5.0   
3     {'$oid': '5ff1e1ee0a7214ada100056f'}                5.0   
4     {'$oid': '5ff1e1d20a7214ada1000561'}                5.0   
...                                    ...                ...   
1114  {'$oid': '603cc0630a720fde100003e6'}               25.0   
1115  {'$oid': '603d0b710a720fde1000042a'}                NaN   
1116  {'$oid': '603cf5290a720fde10000413'}                NaN   
1117  {'$oid': '603ce7100a7217c72c000405'}               25.0   
1118  {'$oid': '603c4fea0a7217c72c000389'}                NaN   

                                bonusPointsEarnedReason  \
0     Receipt number 2 completed, bonus point schedu...   
1     Receipt number 5 completed, bonus point schedu...   
2                            All-receipts receipt bonus   
3                            All-receipts receipt bonus   
4                            All-receipts receipt bonus   
...                                                 ...   
1114                        COMPLETE_NONPARTNER_RECEIPT   
1115                                                NaN   
1116                                                NaN   
1117                        COMPLETE_NONPARTNER_RECEIPT   
1118                                                NaN   

                    createDate               dateScanned  \
0     {'$date': 1609687531000}  {'$date': 1609687531000}   
1     {'$date': 1609687483000}  {'$date': 1609687483000}   
2     {'$date': 1609687537000}  {'$date': 1609687537000}   
3     {'$date': 1609687534000}  {'$date': 1609687534000}   
4     {'$date': 1609687506000}  {'$date': 1609687506000}   
...                        ...                       ...   
1114  {'$date': 1614594147000}  {'$date': 1614594147000}   
1115  {'$date': 1614613361873}  {'$date': 1614613361873}   
1116  {'$date': 1614607657664}  {'$date': 1614607657664}   
1117  {'$date': 1614604048000}  {'$date': 1614604048000}   
1118  {'$date': 1614565354962}  {'$date': 1614565354962}   

                  finishedDate                modifyDate  \
0     {'$date': 1609687531000}  {'$date': 1609687536000}   
1     {'$date': 1609687483000}  {'$date': 1609687488000}   
2                          NaN  {'$date': 1609687542000}   
3     {'$date': 1609687534000}  {'$date': 1609687539000}   
4     {'$date': 1609687511000}  {'$date': 1609687511000}   
...                        ...                       ...   
1114                       NaN  {'$date': 1614594148000}   
1115                       NaN  {'$date': 1614613361873}   
1116                       NaN  {'$date': 1614607657664}   
1117                       NaN  {'$date': 1614604049000}   
1118                       NaN  {'$date': 1614565354962}   

             pointsAwardedDate  pointsEarned              purchaseDate  \
0     {'$date': 1609687531000}         500.0  {'$date': 1609632000000}   
1     {'$date': 1609687483000}         150.0  {'$date': 1609601083000}   
2                          NaN           5.0  {'$date': 1609632000000}   
3     {'$date': 1609687534000}           5.0  {'$date': 1609632000000}   
4     {'$date': 1609687506000}           5.0  {'$date': 1609601106000}   
...                        ...           ...                       ...   
1114                       NaN          25.0  {'$date': 1597622400000}   
1115                       NaN           NaN                       NaN   
1116                       NaN           NaN                       NaN   
1117                       NaN          25.0  {'$date': 1597622400000}   
1118                       NaN           NaN                       NaN   

      purchasedItemCount                             rewardsReceiptItemList  \
0                    5.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
1                    2.0  [{'barcode': '4011', 'description'

## Lets perform some preprocessing 
### Changing column names


In [165]:
users = users.rename(columns={'_id':'userId'})
brands = brands.rename(columns={"_id": "brandId"})
receipts = receipts.rename(columns={'_id':'receiptId'})
users

userId  active               createdDate  \
0    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
1    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
3    {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
4    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
..                                    ...     ...                       ...   
490  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
491  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
492  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
493  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
494  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   

                    lastLogin         role signUpSource state  
0    {'$date': 1609687537858}     consumer        Email    WI  
1    {'$date': 1609687537858}     consumer        Email    WI  
2    {'$date': 1609687537858}     consumer        Email    WI  
3    {'$date': 1609687530597}     consumer        Email    WI  
4    {'$date': 1609687537858}     consumer        Email    WI  
..                        ...          ...          ...   ...  
490  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
491  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
492  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
493  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
494  {'$date': 1614963143204}  fetch-staff          NaN   NaN  

[495 rows x 7 columns]

### Some of the records are nested, lets correct this by getting the original values present


In [166]:
def replace_dict_with_oid_or_date(df):
    for col in df.columns:
        if df[col].apply(lambda x: isinstance(x, dict)).any():
            first_dict = df[col].iloc[0]
            if any(key.startswith('$oid') or key.startswith('$date') for key in first_dict.keys()):
                df[col] = df[col].apply(lambda x: x[list(filter(lambda k: k.startswith('$oid') or k.startswith('$date'), x.keys()))[0]] if isinstance(x, dict) else x)
    return df
users = replace_dict_with_oid_or_date(users)
brands = replace_dict_with_oid_or_date(brands)
receipts = replace_dict_with_oid_or_date(receipts)
users

userId  active    createdDate     lastLogin  \
0    5ff1e194b6a9d73a3a9f1052    True  1609687444800  1.609688e+12   
1    5ff1e194b6a9d73a3a9f1052    True  1609687444800  1.609688e+12   
2    5ff1e194b6a9d73a3a9f1052    True  1609687444800  1.609688e+12   
3    5ff1e1eacfcf6c399c274ae6    True  1609687530554  1.609688e+12   
4    5ff1e194b6a9d73a3a9f1052    True  1609687444800  1.609688e+12   
..                        ...     ...            ...           ...   
490  54943462e4b07e684157a532    True  1418998882381  1.614963e+12   
491  54943462e4b07e684157a532    True  1418998882381  1.614963e+12   
492  54943462e4b07e684157a532    True  1418998882381  1.614963e+12   
493  54943462e4b07e684157a532    True  1418998882381  1.614963e+12   
494  54943462e4b07e684157a532    True  1418998882381  1.614963e+12   

            role signUpSource state  
0       consumer        Email    WI  
1       consumer        Email    WI  
2       consumer        Email    WI  
3       consumer        Email    WI  
4       consumer        Email    WI  
..           ...          ...   ...  
490  fetch-staff          NaN   NaN  
491  fetch-staff          NaN   NaN  
492  fetch-staff          NaN   NaN  
493  fetch-staff          NaN   NaN  
494  fetch-staff          NaN   NaN  

[495 rows x 7 columns]

##

## Creating new tables

In [167]:
## CPG TABLE
def extract_values(collection_dict):
    if isinstance(collection_dict, dict):
        oid = collection_dict['$id']['$oid']
        ref = collection_dict['$ref']
        return pd.Series({'oid': oid, 'ref': ref})
    else:
        return pd.Series({'oid': None, 'ref': None})
cpg = brands['cpg'].apply(extract_values)
brands.drop('cpg',axis=1,inplace=True)
cpg

oid   ref
0     601ac114be37ce2ead437550  Cogs
1     5332f5fbe4b03c9a25efd0ba  Cogs
2     601ac142be37ce2ead437559  Cogs
3     601ac142be37ce2ead437559  Cogs
4     5332fa12e4b03c9a25efd1e7  Cogs
...                        ...   ...
1162  5f77274dbe37ce6b592e90bf  Cogs
1163  53e10d6368abd3c7065097cc  Cogs
1164  5332fa12e4b03c9a25efd1e7  Cogs
1165  5332f5f6e4b03c9a25efd0b4  Cogs
1166  6026d757be37ce6369301467  Cogs

[1167 rows x 2 columns]

### Retrieving nested information from the receipts file

In [169]:

def explode_dict_lists(df, col_name):
    all_keys = set()
    for record in df[col_name]:
        if isinstance(record, list):
            for item in record:
                if isinstance(item, dict):
                    all_keys.update(item.keys())
    new_df = pd.DataFrame(columns=sorted(all_keys))
    
    for idx, row in df.iterrows():
        record = row[col_name]
        id_value = row['receiptId']
        if isinstance(record, list):
            new_row = pd.Series({key: None for key in all_keys}, index=all_keys)
            for item in record:
                item.update({'receiptId':id_value})
                if isinstance(item, dict):
                    for key, value in item.items():
                        new_row[key] = value
                    new_row['receiptId'] = id_value
                    new_df.loc[len(new_df)] = new_row
        else:
        
            new_row = pd.Series({key: None for key in all_keys}, index=all_keys)
            new_row['receiptId'] = id_value
            new_df.loc[len(new_df)] = new_row

    return new_df
rewardsReceiptItemInfo= explode_dict_lists(receipts,'rewardsReceiptItemList')
receipts.drop(['rewardsReceiptItemList'],axis=1,inplace=True)
rewardsReceiptItemInfo

barcode brandCode competitiveProduct competitorRewardsGroup  \
0             4011      None               None                   None   
1             4011      None               None                   None   
2     028400642255      None               None                   None   
3             None      None               None                   None   
4             4011      None               None                   None   
...            ...       ...                ...                    ...   
7376          None      None               None                   None   
7377          None      None               None                   None   
7378    B076FJ92M4      None               None                   None   
7379    B07BRRLSVC      None               None                   None   
7380          None      None               None                   None   

     deleted                                        description  \
0       None                                     ITEM NOT FOUND   
1       None                                     ITEM NOT FOUND   
2       None  DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCE...   
3       None                                               None   
4       None                                     ITEM NOT FOUND   
...      ...                                                ...   
7376    None                                               None   
7377    None                                               None   
7378    None  mueller austria hypergrind precision electric ...   
7379    None  thindust summer face mask - sun protection nec...   
7380    None                                               None   

     discountedItemPrice finalPrice itemNumber itemPrice  ...  \
0                   None      26.00       None     26.00  ...   
1                   None          1       None         1  ...   
2                   None      10.00       None     10.00  ...   
3                   None       None       None      None  ...   
4                   None      28.00       None     28.00  ...   
...                  ...        ...        ...       ...  ...   
7376                None       None       None      None  ...   
7377                None       None       None      None  ...   
7378               22.97      22.97       None     22.97  ...   
7379               11.99      11.99       None     11.99  ...   
7380                None       None       None      None  ...   

     quantityPurchased                 receiptId  \
0                    5  5ff1e1eb0a720f0523000575   
1                    1  5ff1e1bb0a720f052300056b   
2                    1  5ff1e1bb0a720f052300056b   
3                 None  5ff1e1f10a720f052300057a   
4                    4  5ff1e1ee0a7214ada100056f   
...                ...                       ...   
7376              None  603d0b710a720fde1000042a   
7377              None  603cf5290a720fde10000413   
7378                 1  603ce7100a7217c72c000405   
7379                 1  603ce7100a7217c72c000405   
7380              None  603c4fea0a7217c72c000389   

                                rewardsGroup   rewardsProductPartnerId  \
0                                       None                      None   
1                                       None                      None   
2     DORITOS SPICY SWEET CHILI SINGLE SERVE  5332f5fbe4b03c9a25efd0ba   
3                                       None                      None   
4                                       None                      None   
...                                      ...                       ...   
7376                                    None                      None   
7377                                    None                      None   
7378                                    None                      None   
7379                                    None                      None   
7380                                    None                      None   

     targetPrice userF

### Using SQLite flavor

In [61]:
from pandasql import sqldf
pysqldf = lambda q: sqldf(q, globals())


## Answering Business Questions using SQL Queries

### Q1. What are the top 5 brands by receipts scanned for most recent month?

#### There are two ways this question can be interpreted:
#### 1. For the month of the latest date scanned, get the top 5 brands (if the last date was July 3 2021, we would get all records for july 2021)
#### 2. Get the top 5 brands from a period of 1 month from the latest date scanned (for eg if latest date scanned was July 11 2021, this would mean records from June 11 to July 11)

#### I will follow the second approach as it is the common way of how companies store data 

In [217]:
query = """ with most_recent_month_sales as ( SELECT receiptId
FROM receipts
WHERE dateScanned >= ((SELECT MAX(dateScanned) FROM receipts) - (31 * 24 * 60 * 60 * 1000))
)

SELECT brandCode,COUNT(*) as brand_count
FROM receipts rec
INNER JOIN rewardsReceiptItemInfo re ON rec.receiptId = re.receiptId
WHERE rec.receiptId IN (SELECT receiptId FROM most_recent_month_sales)
GROUP BY brandCode
ORDER BY brand_count DESC


  """

result = pysqldf(query)
result

brandCode  brand_count
0      None          612
1     BRAND            3
2   MISSION            2
3      VIVA            1

#### It looks like the brandcode column is mostly null for the last month period from the receipts file. There are only 3 other brands namely BRAND,MISSION,VIVA
#### This seems to be an error that likely occured during data collection or during data entry

### Q2. How does the ranking of the top 5 brands by receipts scanned for the recent month compare to the ranking for the previous month?

In [220]:
query = """ with most_recent_month_sales as ( SELECT receiptId
FROM receipts
WHERE dateScanned <= (SELECT MAX(dateScanned) FROM receipts) - (31 * 24 * 60 * 60 * 1000)
and dateScanned >= (SELECT MAX(dateScanned) FROM receipts) - (62 * 24 * 60 * 60 * 1000)
)

SELECT brandCode,COUNT(*) as brand_count
FROM receipts rec
INNER JOIN rewardsReceiptItemInfo re ON rec.receiptId = re.receiptId
WHERE rec.receiptId IN (SELECT receiptId FROM most_recent_month_sales)
GROUP BY brandCode
ORDER BY brand_count DESC
LIMIT 6


  """

result = pysqldf(query)
result

brandCode  brand_count
0            None         1424
1  BEN AND JERRYS          406
2          HY-VEE          337
3          KROGER          179
4          BORDEN          154
5           PEPSI          152

#### Although there is a lot of missing data in the previous month (None value counts are the highest again), we can see a healthier and denser distribution of brands
#### The TOP 5 brands in the previous month are Ben and Jerry's , Hyvee, Kroger, Borden and Pepsi

### Q3. When considering average spend from receipts with 'rewardsReceiptStatus’ of ‘Accepted’ or ‘Rejected’, which is greater?

In [86]:
query = """
        SELECT distinct(rewardsReceiptStatus)
        FROM receipts
        """

result = pysqldf(query)
result

rewardsReceiptStatus
0             FINISHED
1             REJECTED
2              FLAGGED
3            SUBMITTED
4              PENDING

#### Upon simply looking at the data, there is no status such as ACCEPTED for the receipts anywhere, None of the other values match closely in meaning except for FINISHED,
#### We will assume that FINISHED status has the same meaning as ACCEPTED, although it could mean something else entirely too, ideally this issue should be checked for and solved as early as possible 

In [96]:
query = """
        SELECT AVG(totalSpent),rewardsReceiptStatus
        FROM receipts
        WHERE rewardsReceiptStatus IN ('FINISHED','REJECTED')
        GROUP BY rewardsReceiptStatus
        """


# Execute the SQL query
result = pysqldf(query)
result

AVG(totalSpent) rewardsReceiptStatus
0        80.854305             FINISHED
1        23.326056             REJECTED

#### Receipts with status of FINISHED/ACCEPTED have a higher spend (80.85) as compared to receipts with a REJECTED status (23.32)

### Q4. When considering total number of items purchased from receipts with 'rewardsReceiptStatus’ of ‘Accepted’ or ‘Rejected’, which is greater?

In [97]:
query = """
        SELECT SUM(purchasedItemCount),rewardsReceiptStatus
        FROM receipts
        WHERE rewardsReceiptStatus IN ('FINISHED','REJECTED')
        GROUP BY rewardsReceiptStatus
        """
result = pysqldf(query)
result

SUM(purchasedItemCount) rewardsReceiptStatus
0                   8184.0             FINISHED
1                    173.0             REJECTED

#### Receipts with FINISHED status have more total number of items purchased (8184) as compared to receipts with a REJECTED status (173)

### Q5. Which brand has the most spend among users who were created within the past 6 months?

In [292]:
query = """ WITH six_month_user_ids AS (
SELECT distinct(userId)
FROM users
WHERE createdDate >= (SELECT MAX(createdDate) FROM users) - (186 * 24 * 60 * 60 * 1000)
order by createdDate
)

select userId,sum(totalSpent) as tot,rew.brandCode from (select * from six_month_user_ids su inner join receipts re on su.userId = re.userId) as meow
inner join rewardsReceiptItemInfo rew on rew.receiptId = meow.receiptId
group by rew.brandCode
order by tot desc
LIMIT 1


  """

result = pysqldf(query)
result

userId        tot       brandCode
0  60023de5fb296c121a81b955  526752.69  BEN AND JERRYS

#### Ben and Jerrys has the highest spend across users who created an account in the last six months

### Q6. Which brand has the most transactions among users who were created within the past 6 months?

In [267]:
query = """ WITH six_month_user_ids AS (
SELECT distinct(userId)
FROM users
WHERE createdDate >= (SELECT MAX(createdDate) FROM users) - (186 * 24 * 60 * 60 * 1000)
order by createdDate
)

select count(rew.brandCode) as transactions,rew.brandCode from (select * from six_month_user_ids su inner join receipts re on su.userId = re.userId) as meow
inner join rewardsReceiptItemInfo rew on rew.receiptId = meow.receiptId
group by rew.brandCode
order by transactions

  """

result = pysqldf(query)
result

transcations       brandCode
0               0            None
1               1          ARNOLD
2               1        BOTA BOX
3               1       CARAMELLO
4               1            DARE
..            ...             ...
163           122           PEPSI
164           124         KLEENEX
165           129            DOLE
166           218  BEN AND JERRYS
167           337          HY-VEE

[168 rows x 2 columns]

#### HY-VEE has the most transactions among users who created an account in the last six months

## Data Quality, Issues and Integrity

#### Right from the get-go, the data provided had a lot of problems with messy strucuture, nested data, missing values

#### I had to clean up the data to start working with them and understand them
#### This was super evident in the original receipts file as there was lot of nested data in a column which was also inconsistent across the rows
#### The users and brands file also contained nested data for which I wrote functions to clean them up

### Missing Data

In [275]:
def percentage_missing_data(df):
    missing_data = df.isnull().sum()
    percentage_missing = (missing_data / df.shape[0]) * 100
    return percentage_missing
percentage_missing_data(users)

userId           0.000000
active           0.000000
createdDate      0.000000
lastLogin       12.525253
role             0.000000
signUpSource     9.696970
state           11.313131
dtype: float64

In [276]:
percentage_missing_data(brands)

brandId          0.000000
barcode          0.000000
category        13.281919
categoryCode    55.698372
name             0.000000
topBrand        52.442159
brandCode       20.051414
dtype: float64

In [277]:
percentage_missing_data(receipts)

receiptId                   0.000000
bonusPointsEarned          51.385165
bonusPointsEarnedReason    51.385165
createDate                  0.000000
dateScanned                 0.000000
finishedDate               49.240393
modifyDate                  0.000000
pointsAwardedDate          52.010724
pointsEarned               45.576408
purchaseDate               40.035746
purchasedItemCount         43.252904
rewardsReceiptStatus        0.000000
totalSpent                 38.873995
userId                      0.000000
dtype: float64

### The brands and receipts files contain a lot of missing data in a majority of the columns; this can hinder future analytical work and wouldnt give the right insights about our data; the users file contains a small portion of missing data, this can be either imputed or the rows can be dropped entirely owing to the small size

### Duplicate rows

In [278]:
users[users.duplicated()]

userId  active    createdDate     lastLogin  \
1    5ff1e194b6a9d73a3a9f1052    True  1609687444800  1.609688e+12   
2    5ff1e194b6a9d73a3a9f1052    True  1609687444800  1.609688e+12   
4    5ff1e194b6a9d73a3a9f1052    True  1609687444800  1.609688e+12   
5    5ff1e194b6a9d73a3a9f1052    True  1609687444800  1.609688e+12   
8    5ff1e194b6a9d73a3a9f1052    True  1609687444800  1.609688e+12   
..                        ...     ...            ...           ...   
490  54943462e4b07e684157a532    True  1418998882381  1.614963e+12   
491  54943462e4b07e684157a532    True  1418998882381  1.614963e+12   
492  54943462e4b07e684157a532    True  1418998882381  1.614963e+12   
493  54943462e4b07e684157a532    True  1418998882381  1.614963e+12   
494  54943462e4b07e684157a532    True  1418998882381  1.614963e+12   

            role signUpSource state  
1       consumer        Email    WI  
2       consumer        Email    WI  
4       consumer        Email    WI  
5       consumer        Email    WI  
8       consumer        Email    WI  
..           ...          ...   ...  
490  fetch-staff          NaN   NaN  
491  fetch-staff          NaN   NaN  
492  fetch-staff          NaN   NaN  
493  fetch-staff          NaN   NaN  
494  fetch-staff          NaN   NaN  

[283 rows x 7 columns]

In [279]:
brands[brands.duplicated()]

Empty DataFrame
Columns: [brandId, barcode, category, categoryCode, name, topBrand, brandCode]
Index: []

In [280]:
receipts[receipts.duplicated()]

Empty DataFrame
Columns: [receiptId, bonusPointsEarned, bonusPointsEarnedReason, createDate, dateScanned, finishedDate, modifyDate, pointsAwardedDate, pointsEarned, purchaseDate, purchasedItemCount, rewardsReceiptStatus, totalSpent, userId]
Index: []

### More than half of the users dataset is duplicated, while the brands and receipts dataset do not contain any duplicates; duplicate data is redundant and can be removed carefree

### Other issues

#### The receipts file orginally contained a lot of nested data in the rewardsReceiptItemList column; This was broken down into a new table to make it easier to acess that data to conduct analysis.


The file also had data inconsistencies with receiptstatus and number of points earned, this is especially unclear in the cases of rejected receipts which also ended up earning reward points. This data inconsistency must be clarified before data is made available for analysis.

In [287]:
receipts[receipts.rewardsReceiptStatus == 'REJECTED'][['pointsAwardedDate','pointsEarned']]

pointsAwardedDate  pointsEarned
2                   NaN           5.0
13         1.604089e+12         750.0
28                  NaN           NaN
62                  NaN         750.0
175                 NaN           NaN
...                 ...           ...
1106                NaN          25.0
1112                NaN          25.0
1113                NaN          25.0
1114                NaN          25.0
1117                NaN          25.0

[71 rows x 2 columns]

In [295]:
## date is in unix timestamp format
query = """
SELECT datetime((createdDate / 1000), 'unixepoch') AS latest_date
FROM users
WHERE createdDate >= (SELECT MAX(createdDate) FROM users) - (350 * 24 * 60 * 60 * 1000)
order by createdDate


        """
result = pysqldf(query)
result



latest_date
0    2020-07-28 18:04:00
1    2020-11-04 22:29:33
2    2020-11-05 15:17:09
3    2020-11-05 15:17:09
4    2020-11-05 15:17:09
..                   ...
441  2021-02-11 16:17:07
442  2021-02-11 16:17:49
443  2021-02-11 16:17:53
444  2021-02-12 14:10:49
445  2021-02-12 14:11:06

[446 rows x 1 columns]

### there have been no users created after July 28 2020 until Nov 4 2020, although this might be normal it is definitely interesting

### There might be other data inconsistencies which can be found easily if there was acess to more comprehensive documentation and metadata

# Recommendations

### 1. Missing values are present everywhere across the three files. The users dataset also had lot of duplicate rows. Proper data quality checks must be implemented in the data pipeline ideally before data loading and ingestion. This can be automated but occasional human interference might be required as data is never perfect everytime.
### 2. Metadata about data files can be vital for data management, integrity and security. There was little information about what the files contain,where they are sourced from etc. I would recommend providing metadata as they also provide background information and context which can be helpful during data analysis.